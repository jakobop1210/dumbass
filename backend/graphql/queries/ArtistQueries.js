const { ArtistType } = require('../types/ArtistType')
const {
    GraphQLID,
    GraphQLList,
    GraphQLNonNull,
    GraphQLInt,
    GraphQLString
} = require('graphql')
const Artist = require('../../models/Artist')

/**
 * GraphQL Query Field - getArtistById.
 *
 * Retrieves a single artist by their unique ID. Returns an `ArtistType` object corresponding to the provided ID.
 *
 * @type {Object} GraphQL query configuration object.
 * @property {ArtistType} type - The GraphQL type that this query will return.
 * @property {Object} args - Arguments required for this query.
 * @property {function} resolve - The resolver function to execute the query.
 */
const getArtistById = {
    type: ArtistType,
    args: {
        id: { type: GraphQLID }
    },
    resolve: async (parent, args) => {
        try {
            const artist = await Artist.findOne({ id: Number(args.id) })
            if (!artist) {
                throw new Error(`Artist with id ${args.id} not found.`)
            }
            return artist
        } catch (error) {
            throw new Error(error.message)
        }
    }
}

/**
 * GraphQL Query Field - getArtistsByIds.
 *
 * Retrieves multiple artists based on an array of their IDs. Returns a list of `ArtistType` objects.
 *
 * @type {Object} GraphQL query configuration object.
 * @property {GraphQLList(ArtistType)} type - The GraphQL type that this query will return.
 * @property {Object} args - Arguments required for this query.
 * @property {function} resolve - The resolver function to execute the query.
 */
const getArtistsByIds = {
    type: new GraphQLList(ArtistType),
    args: {
        ids: { type: new GraphQLNonNull(new GraphQLList(GraphQLInt)) }
    },
    resolve: async (parent, args) => {
        try {
            const artists = await Artist.find({
                id: { $in: args.ids }
            })
            if (!artists) {
                throw new Error(`Artists with ids ${args.ids} not found.`)
            }
            return artists
        } catch (error) {
            throw new Error(error.message)
        }
    }
}

/**
 * GraphQL Query Field - countArtists.
 *
 * Counts the number of artists that match a given criteria, such as a name pattern. Returns an integer count.
 *
 * @type {Object} GraphQL query configuration object.
 * @property {GraphQLInt} type - The GraphQL type that this query will return (integer in this case).
 * @property {Object} args - Arguments required for this query.
 * @property {function} resolve - The resolver function to execute the query.
 */
const countArtists = {
    type: GraphQLInt,
    args: {
        name: { type: GraphQLString }
    },
    resolve: async (parent, args) => {
        let query = {}
        if (args.name) {
            query.name = new RegExp(args.name, 'i')
        }
        try {
            return await Artist.countDocuments(query)
        } catch (error) {
            throw new Error('Error counting artists. ' + error)
        }
    }
}

module.exports = { getArtistById, getArtistsByIds, countArtists }
