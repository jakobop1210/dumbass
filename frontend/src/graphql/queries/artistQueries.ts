import { gql } from '@apollo/client'

export const GET_TOP_ARTISTS = gql`
    query ($limit: Int) {
        getTopArtists(limit: $limit) {
            id
            name
            alternate_names
            description
            image_url
            average_rating
            number_of_ratings
        }
    }
`

export const GET_ARTIST_BY_ID = gql`
    query ($id: ID!) {
        getArtistById(id: $id) {
            alternate_names
            description
            image_url
            name
            average_rating
            number_of_ratings
        }
    }
`

export const GET_ARTISTS_BY_ID = gql`
    query ($ids: [ID!]!) {
        getArtistsByIds(ids: $ids) {
            id
            name
            alternate_names
            description
            image_url
            average_rating
            number_of_ratings
        }
    }
`
