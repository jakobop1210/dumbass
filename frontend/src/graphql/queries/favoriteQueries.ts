import { gql } from '@apollo/client'

/**
 * GraphQL query: CHECK_IF_FAVORITE.
 *
 * This query checks if a specific item (song or artist) is marked as a favorite by a user.
 * It requires the username of the user, the type of the item (song or artist), and the item's ID.
 *
 * @param {string} $username - The username of the user.
 * @param {string} $type - The type of the item ('song' or 'artist').
 * @param {number} $targetId - The ID of the item (song or artist).
 * @returns {boolean} A boolean indicating whether the item is a favorite of the user.
 */
export const CHECK_IF_FAVORITE = gql`
    query ($username: String!, $type: String!, $targetId: Int!) {
        checkIfFavorite(username: $username, type: $type, targetId: $targetId)
    }
`

/**
 * GraphQL query: GET_FAVORITES.
 *
 * This query retrieves a list of favorite items (songs or artists) for a specific user.
 * It requires the username of the user and returns an array of objects, each representing a favorite item.
 * Each object includes the type of the item (song or artist) and its ID.
 *
 * @param {string} $username - The username of the user whose favorites are being requested.
 * @returns {Array<Object>} An array of objects, each containing the type ('song' or 'artist') and the targetId of the favorite item.
 */
export const GET_FAVORITES = gql`
    query ($username: String!) {
        getFavorites(username: $username) {
            type
            targetId
        }
    }
`
