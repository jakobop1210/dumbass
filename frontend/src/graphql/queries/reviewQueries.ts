import { gql } from '@apollo/client'

/**
 * GraphQL query: GET_REVIEWS_BY_TARGET_ID.
 *
 * This query fetches reviews for a specific target item, identified by its type and ID.
 * The target can be either a song or an artist. The query returns a list of reviews, 
 * each containing the username of the reviewer, the content of the review, and the rating given.
 *
 * @param {string} $targetType - The type of the target for which reviews are being fetched ('song' or 'artist').
 * @param {number} $targetId - The unique identifier of the target (song or artist).
 * @returns {Array<Object>} An array of review objects, each containing the userName, content, and rating of the review.
 */
export const GET_REVIEWS_BY_TARGET_ID = gql`
    query ($targetType: String!, $targetId: Int!) {
        getReviewsByTarget(targetType: $targetType, targetId: $targetId) {
            userName
            content
            rating
        }
    }
`
