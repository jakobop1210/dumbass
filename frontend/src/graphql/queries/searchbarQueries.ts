import { gql } from '@apollo/client'

export const SEARCHBAR_DROPDOWN = gql`
    query ($searchType: String!, $searchString: String!, $limit: Int) {
        searchSearchbar(
            searchType: $searchType
            searchString: $searchString
            limit: $limit
        ) {
            ... on Artist {
                name
                id
            }
            ... on Song {
                title
                id
            }
        }
    }
`
