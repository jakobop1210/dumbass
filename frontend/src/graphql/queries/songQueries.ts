import { gql } from '@apollo/client'

/**
 * GraphQL query: GET_SONG_BY_ID.
 *
 * This query fetches detailed information about a song by its unique ID. 
 * The details include the song's lyrics, header image URL, release date, title, artist names, average rating, and the number of ratings.
 *
 * @param {string} $id - The unique ID of the song to fetch.
 * @returns {Object} An object containing detailed information about the song.
 */
export const GET_SONG_BY_ID = gql`
    query ($id: ID!) {
        getSongById(id: $id) {
            lyrics
            header_image_url
            release_date
            title
            artist_names
            average_rating
            number_of_ratings
        }
    }
`

/**
 * GraphQL query: GET_SONGS_BY_ID.
 *
 * This query retrieves detailed information for a list of songs based on their unique IDs.
 * The details include the song's ID, title, artist names, header image URL, release date, primary artist ID, average rating, and number of ratings.
 *
 * @param {Array<number>} $ids - An array of IDs of the songs to fetch.
 * @returns {Array<Object>} An array of objects, each containing detailed information about a song.
 */
export const GET_SONGS_BY_ID = gql`
    query ($ids: [Int]!) {
        getSongsByIds(ids: $ids) {
            id
            title
            artist_names
            header_image_url
            release_date
            primary_artist_id
            average_rating
            number_of_ratings
        }
    }
`

/**
 * GraphQL query: GET_SONGS_ON_TITLE.
 *
 * This query fetches a list of songs based on a search term, with optional parameters for pagination and sorting.
 * The details returned for each song include the header image URL, ID, title, artist names, average rating, number of ratings, and release date.
 *
 * @param {string} [$title] - The title or part of the title of the songs to search for.
 * @param {number} [$limit] - The maximum number of songs to return.
 * @param {string} [$sort] - The sorting criteria (e.g., 'rating', 'alphabetical').
 * @param {number} [$page] - The page number for pagination.
 * @returns {Array<Object>} An array of objects, each containing information about a song.
 */
export const GET_SONGS_ON_TITLE = gql`
    query ($title: String, $limit: Int, $sort: String, $page: Int) {
        getSongsOnTitle(
            title: $title
            limit: $limit
            sort: $sort
            page: $page
        ) {
            header_image_url
            id
            title
            artist_names
            average_rating
            number_of_ratings
            release_date
        }
    }
`

/**
 * GraphQL query: COUNT_SONGS.
 *
 * This query counts the number of songs in the database that match a given title or search term.
 *
 * @param {string} [$title] - The title or part of the title of the songs to count.
 * @returns {number} The count of songs matching the search criteria.
 */
export const COUNT_SONGS = gql`
    query ($title: String) {
        countSongs(title: $title)
    }
`
