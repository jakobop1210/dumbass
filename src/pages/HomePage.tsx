import { useState } from "react";
import SearchBar from "../components/molecules/SearchBar";
import ArtistSongCard, { ArtistCardProps, SongCardProps } from "../components/molecules/ArtistSongCard";

export default function HomePage() {
    const [filter, setFilter] = useState("Song");
    const FiftycentProps = {
        cardType: "artist",
        imageUrl: "https://www.uka.no/uploads/cache/66/e7/66e75771d31a087bd8754021b203d98c.jpg",
        title: "50 Cent",
        alternateNames: ["Fiddy", "Boo Boo"],
        rating: 4.5,
        numOfRatings: 1000000,
    } as ArtistCardProps;
    const InDaClubProps = {
        cardType: "song",
        imageUrl: "https://upload.wikimedia.org/wikipedia/en/1/12/50_Cent_-_In_Da_Club_-_CD_cover.jpg",
        title: "In Da Club",
        artist: "50 Cent",
        rating: 4.5,
        numOfRatings: 1000000,
        releaseDate: "2003-01-07",
    } as SongCardProps;

    return (
        <div className="bg-[#EFEFEF] min-h-screen">
            <div className="flex items-center justify-center h-16 bg-[#FAF3DD] drop-shadow">
            </div>
            <div className="flex flex-col items-center justify-center w-screen px-12">
                <SearchBar
                    className="w-4/5 mt-10 drop-shadow mb-10"
                    filterOptions={["Song", "Artist"]} 
                    selectedFilter={filter}
                    onFilterChange={(newFilter) => setFilter(newFilter)}
                />
                {/* Songs */}
                <div className="w-3/4">
                    <h1 data-testid="top_song_header" className="text-3xl font-bold mt-10 mb-10 text-blueGray self-start">Top Songs</h1>
                    {/* Rendering 6 cards */}
                    <div className="mb-20 grid grid-cols-1 md:grid-cols-1 lg:grid-cols-2 xl:grid-cols-2 2xl:grid-cols-3 sm:gap-8 gap-4">
                        {Array(6).fill(InDaClubProps).map((props, index) => (
                            <ArtistSongCard key={index} {...props} />
                        ))}
                    </div>
                </div>
                
                {/* Artists */}
                <div className="w-3/4">
                    <h1 data-testid="top_artist_header" className="text-3xl font-bold mt-10 mb-10 text-blueGray self-start">Top Artists</h1>
                    {/* Rendering 6 cards */}
                    <div className="mb-20 grid grid-cols-1 md:grid-cols-1 lg:grid-cols-2 xl:grid-cols-2 2xl:grid-cols-3 sm:gap-8 gap-4">
                        {Array(6).fill(FiftycentProps).map((props, index) => (
                            <ArtistSongCard key={index} {...props} />
                        ))}
                    </div>
                </div>
            </div>
        </div>
    );
}
