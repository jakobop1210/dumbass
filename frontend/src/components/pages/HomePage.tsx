import { useState } from 'react'
import { ArtistCardProps, SongCardProps } from '../molecules/ArtistSongCard'
import CardView from '../organisms/CardView'
import CommonSearchBar from '../molecules/CommonSearchBar'
import { GET_TOP_ARTISTS } from '../../graphql/queries/artistQueries'
import { GET_TOP_SONGS } from '../../graphql/queries/songQueries'
import { useQuery } from '@apollo/client'
import { customToast } from '../../lib/utils'
import { ClipLoader } from 'react-spinners'

// Define a type for an artist
type Artist = {
    id: string
    name: string
    image_url: string
    alternate_names: string[]
    average_rating: number
    number_of_ratings: number
}

// Define a type for a song
type Song = {
    id: string
    title: string
    artist_names: string
    header_image_url: string
    release_date: string
    primary_artist_id: number
    average_rating: number
    number_of_ratings: number
}

/**
 * @component ExplorePage
 *
 * `ExplorePage` is a React functional component that displays a page where users
 * can explore various artists and songs using a search bar and view them via card layouts.
 */
export default function HomePage() {
    const [filter, setFilter] = useState('Song')
    const limitData = 12

    // Define GraphQL queries using the useQuery hook
    const {
        loading: loadingArtists,
        error: errorArtists,
        data: dataArtists
    } = useQuery(GET_TOP_ARTISTS, {
        variables: { limit: limitData }
    })
    const {
        loading: loadingSongs,
        error: errorSongs,
        data: dataSongs
    } = useQuery(GET_TOP_SONGS, {
        variables: { limit: limitData }
    })

    if (errorArtists || errorSongs) {
        console.log(
            `Error: ${JSON.stringify(errorArtists || errorSongs, null, 2)}`
        )

        customToast('error', 'Error', 'Could not load data')
    }

    // Combine and display data
    const artists: Artist[] = dataArtists?.getTopArtists || []
    const songs: Song[] = dataSongs?.getTopSongs || []

    const artistCardData: ArtistCardProps[] = artists.map((artist: Artist) => ({
        cardType: 'artist',
        id: artist.id,
        title: artist.name,
        alternateNames: artist.alternate_names,
        imageUrl: artist.image_url,
        rating: artist.average_rating,
        numOfRatings: artist.number_of_ratings
    }))

    const songCardData: SongCardProps[] = songs.map((song: Song) => ({
        cardType: 'song',
        id: song.id,
        title: song.title,
        artist: song.artist_names,
        imageUrl: song.header_image_url,
        rating: song.average_rating,
        numOfRatings: song.number_of_ratings,
        releaseDate: song.release_date
    }))

    return (
        <main className='flex flex-col items-center justify-center w-screen'>
            <CommonSearchBar
                className='w-4/5 mt-10 drop-shadow mb-10'
                filterOptions={['Song', 'Artist']}
                selectedFilter={filter}
                onFilterChange={(newFilter) => setFilter(newFilter)}
            />
            {loadingArtists || loadingSongs ? (
                <div className='h-96 flex items-center'>
                    <ClipLoader color={'#8fc0a9'} size={100} />
                </div>
            ) : (
                <>
                    {errorArtists || errorSongs ? (
                        <div className='h-96 flex items-center'>
                            <h1 className='text-blueGray text-2xl'>
                                Error loading data :/
                            </h1>
                        </div>
                    ) : (
                        <>
                            <section className='w-full flex justify-center'>
                                <CardView
                                    title='Top Songs'
                                    cardData={songCardData}
                                />
                            </section>
                            <section className='w-full flex justify-center'>
                                <CardView
                                    title='Top Artists'
                                    cardData={artistCardData}
                                />
                            </section>
                        </>
                    )}
                </>
            )}
        </main>
    )
}
