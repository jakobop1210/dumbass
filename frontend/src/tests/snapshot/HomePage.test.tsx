import { render } from '@testing-library/react'
import { BrowserRouter as Router } from 'react-router-dom'
import { MockedProvider } from '@apollo/client/testing'
import HomePage from '../../components/pages/HomePage'
import { GET_ARTISTS_ON_NAME } from '../../graphql/queries/artistQueries'
import { GET_SONGS_ON_TITLE } from '../../graphql/queries/songQueries'

const mocks = [
    {
        request: {
            query: GET_ARTISTS_ON_NAME,
            variables: { limit: 12, sort: 'rating', page: 1 }
        },
        result: {
            data: {
                getArtistsOnName: [
                    {
                        id: '1',
                        name: 'Artist 1',
                        alternate_names: ['A1'],
                        image_url: 'http://example.com/artist1.jpg',
                        average_rating: 4.5,
                        number_of_ratings: 100
                    }
                ]
            }
        }
    },
    {
        request: {
            query: GET_SONGS_ON_TITLE,
            variables: { limit: 12, sort: 'rating', page: 1 }
        },
        result: {
            data: {
                getSongsOnTitle: [
                    {
                        header_image_url: 'http://example.com/song1.jpg',
                        id: '1',
                        title: 'Song 1',
                        artist_names: ['Artist 1', 'Artist 2'],
                        average_rating: 4.7,
                        number_of_ratings: 150,
                        release_date: '2021-01-01'
                    }
                ]
            }
        }
    }
]

test('HomePage snapshot', () => {
    const { asFragment } = render(
        <MockedProvider mocks={mocks} addTypename={false}>
            <Router>
                <HomePage />
            </Router>
        </MockedProvider>
    )
    expect(asFragment()).toMatchSnapshot()
})
