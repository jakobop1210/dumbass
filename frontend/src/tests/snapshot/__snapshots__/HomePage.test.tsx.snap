// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomePage snapshot 1`] = `
<DocumentFragment>
  <main
    class="flex flex-col items-center justify-center w-screen"
  >
    <div
      class="relative w-11/12 sm:w-3/5 mt-10 drop-shadow mb-5 z-40"
    >
      <h3
        class="hidden md:block text-xs m-1 opacity-70"
      >
        Enter your search term and press 'Enter' to search
      </h3>
      <form
        autocomplete="off"
        class="flex items-center pl-2 bg-[#FFFFFF] rounded-lg h-14"
        role="search"
      >
        <label
          class="sr-only"
          for="searchInput"
        >
          Search for a song or artist
        </label>
        <input
          class="w-full p-2 outline-none rounded-md"
          id="searchInput"
          placeholder="Search for a song or artist"
          type="text"
          value=""
        />
        <div
          class="h-full border-x-[1.5px] flex justify-center items-center"
        >
          <div
            class="relative inline-block text-left h-full"
            data-headlessui-state=""
          >
            <div
              class="absolute -top-5 text-xs mb-1 opacity-70"
              role="button"
            >
              Filter by
            </div>
            <section
              class="h-full w-full"
            >
              <button
                aria-expanded="false"
                aria-haspopup="menu"
                class="w-24 inline-flex justify-center items-center h-full px-4 py-2 bg-white text-sm font-medium text-gray-700 hover:bg-gray-50  focus-visible:ring-2 focus-visible:ring-white transition-all focus-visible:ring-opacity-75"
                data-headlessui-state=""
                id="headlessui-menu-button-:r0:"
                type="button"
              >
                <span
                  class="block overflow-hidden text-ellipsis whitespace-nowrap"
                >
                  Song
                </span>
                <svg
                  aria-hidden="true"
                  class="ml-2 -mr-1 h-5 w-5"
                  fill="currentColor"
                  viewBox="0 0 20 20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </section>
          </div>
        </div>
        <button
          aria-label="Search"
          class="p-2 rounded-md hover:bg-gray-50 transition-all h-full w-16 flex justify-center items-center rounded-l-none"
          data-testid="search-button"
          type="submit"
        >
          <svg
            color="#999"
            data-testid="search-icon"
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(153, 153, 153);"
            viewBox="0 0 512 512"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
        </button>
      </form>
    </div>
    <div
      class="h-96 flex items-center"
    >
      <span
        style="width: 100px; height: 100px; border-radius: 100%; border: 2px solid; border-top-color: #8fc0a9; border-bottom-color: transparent; border-left-color: #8fc0a9; border-right-color: #8fc0a9; display: inline-block; animation: react-spinners-ClipLoader-clip 0.75s 0s infinite linear; animation-fill-mode: both;"
      />
    </div>
  </main>
</DocumentFragment>
`;
